### PR TITLE
:tada: NEW: CUSTOM_DECORATORS

### PR DESCRIPTION
**Pull request by jycouet**

### CODE REVIEW

File: src/cli.ts [hunk 1]
```diff
@@ -73,13 +80,20 @@ async function main() {
 		);
 		connectionString = answer.connectionString;
 	}
+	const custom_decorators2: Record<string, string> = {
+		"@Fields.string": "@KitFields.string#@kitql/remult",
+	};
+	console.log(`custom_decorators2`, JSON.stringify(custom_decorators2));
+
+	console.log(`custom_decorators`, custom_decorators);
```
Comment (Yedidyar) [inline src/cli.ts:88]: you probably forgot it

File: src/getEntityTypescriptPostgres.test.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
+/* eslint-disable sonarjs/no-duplicate-string */
+import { expect, test, describe } from "vitest";
+import { build_column } from "./getEntityTypescriptPostgres.js";
+
+describe("#unit-test build_column", () => {
+	test("string not nullable wo default val", () => {
+		const info = build_column({
```
Comment (Yedidyar) [inline src/getEntityTypescriptPostgres.test.ts:7]: In general I think that all variables or functions should be camelCase

File: src/getEntityTypescriptPostgres.ts [hunk 3]
```diff
@@ -224,19 +240,35 @@ async function getEntityTypescriptPostgres(
...
+			datetimePrecision,
+			characterMaximumLength,
+			udtName,
+			report,
+			enums,
+			provider,
+			table,
+		});
+
+		// const custom_decorators: Record<string, string> = {
```
Comment (Yedidyar) [inline src/getEntityTypescriptPostgres.ts:261]: remove it

File: src/getEntityTypescriptPostgres.ts [hunk 4]
```diff
@@ -224,19 +240,35 @@ async function getEntityTypescriptPostgres(
...
+		// 	"@Fields.string": "@KitFields.string#@kitql/remult",
+		// };
+
+		let decorator = decorator_infered;
```
Comment (Yedidyar) [inline src/getEntityTypescriptPostgres.ts:265]: maybe should be a find function

File: src/getEntityTypescriptPostgres.ts [hunk 5]
```diff
@@ -334,6 +379,7 @@ const generateEntityString = (
 		`import { Entity, ${
 			isContainsForeignKeys || enumsKeys.length > 0 ? "Field, " : ""
 		}Fields } from 'remult'` +
+		`${addLineIfNeeded([...new Set(additionnalImports)], (c) => `${c}`)}` +
```
Comment (Yedidyar) [inline src/getEntityTypescriptPostgres.ts:382]: `additionnalImports` already a `string[]`

File: src/cli.ts [hunk 6]
```diff
@@ -40,6 +41,12 @@ async function main() {
 				hidden: true,
 				default: process.env["TMP_JYC"] === "true",
 			},
+			custom_decorators: {
+				type: "string",
+				hidden: true,
+				default: process.env["CUSTOM_DECORATORS"] ?? "{}",
+				description: `Example CUSTOM_DECORATORS = '{"@Fields.string":"@KitFields.string#@kitql/remult"}', it will be JSON parsed!`,
```
Comment (Yedidyar) [inline src/cli.ts:48]: i think it needs more docs
for example explain the syntax
why should the exist

File: src/cli.ts [hunk 7]
```diff
@@ -73,13 +80,20 @@ async function main() {
...
+
+	console.log(`custom_decorators`, custom_decorators);
 
 	const spinner = p.spinner();
 	spinner.start("Generating everything for you");
 	const report = await getEntitiesTypescriptPostgres(
 		connectionString,
 		output,
 		tableProps,
+		JSON.parse(custom_decorators),
```
Comment (Yedidyar) [inline src/cli.ts:96]: JSON.parse can error hadle it gracefully

**Review by Yedidyar — CHANGES_REQUESTED** [review]

File: src/cli.ts [hunk 8]
```diff
@@ -21,6 +21,7 @@ async function main() {
 		output,
 		tableProps,
 		tmp_jyc,
+		custom_decorators,
```
Comment (Yedidyar) [inline src/cli.ts:24]:
```suggestion
		tmp-jyc,
		custom-decorators,
```

i think we should rethink multiword arg case
image
image

File: src/getEntityTypescriptPostgres.test.ts (hunk 2 above)
Comment (jycouet) [reply src/getEntityTypescriptPostgres.test.ts:7]: I like snake_case more and more :)
But ok, I will try to no put too much in snake

Comment (Yedidyar) [reply src/getEntityTypescriptPostgres.test.ts:7]: Its just the js standard
You see to be more python guy 🫢

File: src/getEntityTypescriptPostgres.ts (hunk 4 above)
Comment (jycouet) [reply src/getEntityTypescriptPostgres.ts:265]: Found even better 👍

File: src/getEntityTypescriptPostgres.ts [hunk 9]
```diff
@@ -80,14 +93,14 @@ function build_column({
 		current_col += " = " + defaultVal;
 	}
 
-	return current_col;
+	return { col: current_col, decorator_fn, decorator_import };
 }
 
-// TODO: everything optionnal
 export async function getEntitiesTypescriptPostgres(
```
Comment (Yedidyar) [inline src/getEntityTypescriptPostgres.ts:99]: image

**Review by Yedidyar — CHANGES_REQUESTED** [review]

File: src/getEntityTypescriptPostgres.ts (hunk 9 above)
Comment (jycouet) [reply src/getEntityTypescriptPostgres.ts:99]: What is this message? 
I don't have this :o

Comment (Yedidyar) [reply src/getEntityTypescriptPostgres.ts:99]: It coms from Eslint

Comment (jycouet) [reply src/getEntityTypescriptPostgres.ts:99]: I don't have it, and running lint doesn't conplain.

image


I'm missing something?

Comment (Yedidyar) [reply src/getEntityTypescriptPostgres.ts:99]: I dont know why you can't see it

Comment (jycouet) [reply src/getEntityTypescriptPostgres.ts:99]: Because you have a sonar plugin in your IDE?
You have no issue running the lint command as well I guess?

Comment (Yedidyar) [reply src/getEntityTypescriptPostgres.ts:99]: no i disable it and it still errors i add it to ci but it like your machine maybe it linux?

Comment (jycouet) [reply src/getEntityTypescriptPostgres.ts:99]: I'm on windows, running WSL (linux).

When you do: `pnpm lint`, you have an error?

**Comment by Yedidyar** [comment]
[b6f6c0b](https://github.com/Yedidyar/remult-cli/pull/22/commits/b6f6c0b30585eb13bd33eef77eb155c6c7dd7d7b)  I said that it is BREAKING CHANGES but actually all the public flags works fine

File: src/getEntityTypescriptPostgres.ts [hunk 10]
```diff
@@ -258,66 +255,39 @@ async function getEntityTypescriptPostgres(
...
-		}
+		const decorator = customDecorators[decoratorInfered] ?? decoratorInfered;
 
 		// TODO: extract this logic from the process column
 		await handleEnums(enums, dataType, provider, udtName);
 
-		if (!defaultOrderBy && ["order", "name", "nom"].includes(columnName)) {
+		if (
+			!defaultOrderBy &&
+			["order", "name", "nom", "username"].includes(columnName)
```
Comment (Yedidyar) [inline src/getEntityTypescriptPostgres.ts:265]: why "name", "nom", "username" ?

File: src/getEntityTypescriptPostgres.ts [hunk 11]
```diff
@@ -379,7 +376,7 @@ const generateEntityString = (
 		`import { Entity, ${
 			isContainsForeignKeys || enumsKeys.length > 0 ? "Field, " : ""
 		}Fields } from 'remult'` +
-		`${addLineIfNeeded([...new Set(additionnalImports)], (c) => `${c}`)}` +
+		`${addLineIfNeeded([...new Set(additionnalImports)], (c) => c)}` +
```
Comment (Yedidyar) [inline src/getEntityTypescriptPostgres.ts:379]: maybe `addLineIfNeeded` can get an optional format function

**Review by Yedidyar — APPROVED** [review]

File: src/getEntityTypescriptPostgres.ts (hunk 11 above)
Comment (jycouet) [reply src/getEntityTypescriptPostgres.ts:379]: It can. Is it game changer?

Comment (Yedidyar) [reply src/getEntityTypescriptPostgres.ts:379]: No you dont have to do it now it NTH

File: src/getEntityTypescriptPostgres.ts (hunk 10 above)
Comment (jycouet) [reply src/getEntityTypescriptPostgres.ts:265]: It's good default default in my case.

I always keep in mind that it's a one shoot generator.
When my entities will be good, and my app running with remult, I will never use the cli anymore.

Cli is good to Kickstart a project when you have a db. After, I don't the need

**Pull request merged**